File: programs/aiyagari.py

import numpy as np
import math
from scipy.optimize import fminbound, brentq
from lininterp import LinInterp

# Parameters, a la Aiyagari 
w = 1.3712
r =  0.0129
beta = 0.96
mu = 2
sigma = 0.4
m = 1 - mu

m = 1 - mu
def U(c): 
    return ((c + 1e-10)**m - 1) / m

def Up(c): # derivative
    return (c + 1e-10)**(-mu)

# The L shock has 3 states: {smin, 1, smax}, and is uniform.  We
# assume that |smin -1| = |smax - 1| = d, so that E L = 1, and 
# Var L = (2/3) * d**2.  We choose d so that Var L = sigma**2
d = math.sqrt(3.0 / 2.0) * sigma
smin, smax = 1 - d, 1 + d

gridmin, gridmax, gridsize = w * smin, 14, 150
grid = np.linspace(gridmin, gridmax, gridsize)

def maximizer(h, a, b):
    return fminbound(lambda x: -h(x), a, b)

# The objective function for the v-greedy problem
def obj_fun(z, a, v):
    t = v(w * smin + (1 + r) * a) \
                + v(w + (1 + r) * a) + v(w * smax + (1 + r) * a)
    return U(z - a) + (beta / 3) * t

def bellman(v):
    """
    The approximate Bellman operator.
    Parameters: v is a vectorized function (i.e., a callable object which 
        acts pointwise on arrays).
    Returns: An instance of LinInterp.
    """
    vals = []
    for z in grid:
        h = lambda a: obj_fun(z, a, v)
        vals.append(h(maximizer(h, 0, z)))
    return LinInterp(grid, vals)


def find_zb(policy_function):
    h = lambda z: - Up(z) + (beta * (1 + r) / 3) * (\
              Up(w * smin - policy_function(w * smin))  \
            + Up(w - policy_function(w))  \
            + Up(w * smax - policy_function(w * smax)) )
    return brentq(h, gridmin, gridmax)


def compute_optimal_policy(read_from_file=True):

    if read_from_file:
        policy = LinInterp(np.loadtxt('grid_dat.txt'), np.loadtxt('pol_dat.txt'))

    else:
        v = U          # Initial condition
        tol = 0.1      # Error tolerance 

        while 1:
            new_v = bellman(v)
            err = np.max(np.abs(new_v(grid) - v(grid)))
            print err
            if err < tol:
                break
            v = new_v

        policy_on_grid = []
        for z in grid:
            h = lambda a: obj_fun(z, a, v)
            policy_on_grid.append(maximizer(h, 0, z))

        policy = LinInterp(grid, policy_on_grid)

        # for z < zb, set policy(z) = 0.0
        zb = find_zb(policy)
        print "zb = ", zb
        for i in range(gridsize):
            if grid[i] < zb:
                policy_on_grid[i] = 0.0

        np.savetxt('grid_dat.txt', grid, fmt='%.10f')
        np.savetxt('pol_dat.txt', policy_on_grid, fmt='%.10f')

        policy.Y = policy_on_grid

    return policy

File: programs/aiyagari.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <time.h>

#include "utilities.h"
#include "minimize.h"

# define GRIDSIZE 250

/* Parameters */
#define w 1.3712
#define r 0.0129
#define beta 0.96
#define mu 2.0
#define sigma 0.4
/* The L shock has 3 states: {smin, 1, smax}, and is uniform.  We
 * assume that |smin -1| = |smax - 1| = d, so that E L = 1, and 
 * Var L = (2/3) * d**2.  We choose d so that Var L = sigma**2 */
#define d 0.49  /* sqrt(3.0 / 2.0) * sigma  */
#define smin 0.51  /* 1 - d */
#define smax 1.49  /* 1 + d */

/* Some constants derived from the parameters */
double m = 1 - mu;
double betad3 = beta / 3.0;
double R = 1 + r;
double shock_vals[] = {smin, 1, smax};
double gridmin = w * smin;
double gridmax = 14.0;

/* The grid */
double grid[GRIDSIZE]; 
/* First diffs used for linear interpolation */
double grid_diffs[GRIDSIZE-1]; 
/* The approximate optimal policy and successive iterates of the Bellman 
 * operator are represented as piecewise linear functions, with values
 * at the interpotation points stored in the following array */
double vals[GRIDSIZE];  
/* First diffs used for linear interpolation */
double vals_diffs[GRIDSIZE-1];  
double vals_temp[GRIDSIZE];  /* Used when recomputing vals */
/* And the function itself */
double lp(double x) 
{
    return lininterp(grid, vals, grid_diffs, vals_diffs, GRIDSIZE, x);
}

/* Utility function and its first derivative */
double U(double c) 
{
    return (pow(c + 1e-10, m) - 1) / m;
}

double Up(double c) 
{
    return pow(c + 1e-10, -mu);
}

/* The objective function for computing v-greedy policies.  The v 
 * used in the computation is that specified by the current value of
 * vals and vals_diffs.  The argument z is passed as a void pointer
 * in order to create a gsl_function.  The function is multiplied by
 * -1 because we are minimizing. */
double objective(double a, void *params) 
{
    double *z = (double *) params;
    return - U(*z - a) - betad3 * (lp(w * smin + R * a) +
                lp(w + R * a) + lp(w * smax + R * a));
}

void bellman(void) 
{
    int i;
    double result;
    double lower_bound = 0.0;
    double epsilon = 0.000001;
    gsl_function F;
    F.function = &objective;
    for (i = 0; i < GRIDSIZE; i++)
    {
        F.params = &grid[i];
        int exit_val = minimize_convex(&F, lower_bound, grid[i], &result, epsilon);
        vals_temp[i] = - objective(result, &grid[i]);  /* need to take negative */
        lower_bound = result;
    }
    for (i = 0; i < GRIDSIZE; i++)
    {
        vals[i] = vals_temp[i];
    }
}


int main(void)
{
    int i;
    /* Initialize grid and grid_diffs */
    linspace(grid, gridmin, gridmax, GRIDSIZE);
    first_diffs(grid_diffs, grid, GRIDSIZE);
    for (i = 0; i < GRIDSIZE; i++)
    {
        vals[i] = U(grid[i]);
    }
    first_diffs(vals_diffs, vals, GRIDSIZE);
    for (i= 0; i < 20; i++)
    {
        bellman();
    }
    /*
    for (i = 0; i < GRIDSIZE / 2; i++)
    {
        printf("%g %g\n", grid[i], objective(grid[i], &grid[GRIDSIZE / 2]));
    }
    */

    double lower_bound = 0.0;
    double result;
    double epsilon = 0.000001;
    gsl_function F;
    F.function = &objective;
    for (i = 0; i < GRIDSIZE; i++)
    {
        F.params = &grid[i];
        int exit_val = minimize_convex(&F, lower_bound, grid[i], &result, epsilon);
        vals_temp[i] = result;
        lower_bound = result;
    }
    for (i = 0; i < GRIDSIZE; i++)
    {
        vals[i] = vals_temp[i];
    }
    for (i = 0; i < GRIDSIZE; i++)
    {
        printf("%g %g\n", grid[i], w * smax + R * vals[i]);
    }
    return 0;
}
